### ISSUE TITLE
HOC type does not work with Props containing unions of different types

### ISSUE THREAD

**Issue by Blasz** [nordsoftware/react-flow-types#21]
Copying my comment from https://github.com/facebook/flow/issues/5187.

The HOC type doesn't seem to fix the issue for cases where Props is a union of different types.

E.g.
```
type Props = {
  foo: string | number;
}

class myComp extends React.Component { ... }
myComp.defaultProps = {
  foo: ''
};
```

`$Diff` breaks since `Props` and `defaultProps` don't have the same types, see: [this example](https://flow.org/try/#0C4TwDgpgBACgTgezAZygXigbwFBSgMwQQC4plg4BLAOwHMoAfKagVwFsAjCOAbmwF9s2AMYJq5KABEI+AIYsANsHhJUGHHkIkoABgFDQkKZXz50UACSST+ADwqUAGiiGICM9LmLliFAD4+ETEJWVJrU3MNAiJSPX4eIA) for the Diff mismatch case and [this repro case](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAtgBzgJwBdkwBDAZzACUBTUgY2KnzizAHJ87H2BudYMDCFSAaxoA7MM1ZgAFoUI5yALkEBzDITkBXAEYA6eq2AATDJtJwJMDBJrAuDQgFpYiF4QCeOGuXQ0AB54RMI+NGAAwqx49hKEAOpacgAiNFCkOjCEAAosygA8aRlZufmqYADeAL4ANGB5cMoqVdUAfGAAvKhg1NyEBtG41pKEACrhBY3KHQBkVabpmdnTFcXLZU3k1fyooCA94AASNFxgGJSkYOqSpxj0Yb7SBGBHAPKRqoeHYACSxCJxJRCAg4I8IjhSPhSFgaIRThVaABHHQYLimVZkCSmBosABuGEWGPK3xIyNR6NWLTGcgiUDspBgYAQ0JwvhxxmGcWICAwMCZ9hoOPJaKFq3qdjIpnMhAw1mEYK0lEQUhwJJIjQJRKpYBpdIZTJZpDZQrAnNio2ZfIFcGIgpxmsJYvK9T0NHomXIEW0EXekSt-LAargWu9tLYhDBPvOEns+DNMRG8UOaCCIQB4VeFlp+De+EW+CGFviBRFlPKLRq9Ud2orrQ63V68wAFAU8xYDeKwOtSqs2s3zUnCC0i0OktoeytykUlr2XWB25oJIy+wBKLptHq9be0ZyDRPcia+Us0FGi4lbMDzAAkKQwUCgBVv98fi8785rzq2bXqk82MyvMBrwAZTkY0aDbfAO2XGA+zaVctxbSDoJXcp+0HbkWl3Rh9y5UYjwgt8YLXDcwC3bc+j3UdD0mMsv2UQDnwfZCl1QrZq3xJ0LxmeDdlQbwnkxToqjIFoJB0LA3XjAAfMByEIKCJHUV0xIkqSwB2dB6BgChKAAQTAIJ4WxShsIGajRimNCqi3eTSFlB5FhKKdL2EypRLAABGDTdl6LhsVOZt10qciwC4QgdHwKQCnMPEwGATdemqVBkooijQvS7TdKiQzAmM0xTP6XDi0IKzvxsvzJALIKKu3cLIui2L4sSjSUoEIQci9HRTDgFw-WEMDiAAKx0eSg10vwBoiDDLW0FgdHUORUGMCQxuUGhurgI44HoFojmzU48wLCySyresujAZsTuHLEvFXFpSAkLxSJCyqIqiy7g2UddOg6AprqqAwga+7Zmv4ZLUAKAy9E6SpPIABmqMH4qEAA5MFThYfByHqChct8RghUhgzSFhhGkZhuHEeRwQwHRwz8CxnGyEoNN3XhUw9iEKGyE6AAichWAieTFPUPmwEp8maaEABRRmCEuVngnZonlusMaACELvWzbtvoZs9NXfhaeMRn2ZgZ7MYVyGtcl6mEt4FG6Yx+XsdxpWCY5m3eapimyfttpHdp+mrbdln8ZVznaYKLXSYFoW5IUuwxYl-2kYdp25aZ8O2cJzmMu3SH-VJ33U9LjPg5dpn3YjvOubAf6y6livZdd5m8dzjm0sL6Pi-5wXYUT0Xxbt9PA8ztuc+VvPC7V1biBSbWup6vWrqNgvekhxeS6l0fpedhnq6nz2iaKXm+agOA4BHtP96zhWa87oUgA) for the HOC type failing when using union types.

I've tried constraining the DefaultProps generic to `$Shape<Props>` in the `HigherOrderComponent` type but can't seem to get it to work. Probably because it's trying to reference a sibling generic type.

**Comment by Blasz** [comment]
I ended up fixing this by using the [React.ElementConfig](https://flow.org/en/docs/react/hoc/#toc-supporting-defaultprops-with-react-elementconfig) type introduced in 0.63 of flow (and fixed properly in 0.64).

**Comment by cniska** [comment]
@blasz can you make a PR or post your solution here?

**Comment by Blasz** [comment]
@crisu83 I ended up just using React.ElementConfig itself without the HOC type from this project so my solution won't be relevant to this project. I just followed the docs on React.ElementConfig. I only tried using this type as I wasn't aware of the default props fix introduced in 0.64.

Sorry I can't be of more help, I don't mind if this is closed.

**Comment by AriaMinaei** [comment]
According to @rsolomon who contributed to this library, `React.ElementConfig` which was introduced last Jan obviates the need for our `HigherOrderComponent`:  https://github.com/facebook/flow/issues/5187#issuecomment-374995939

Then we should probably put a deprecation warning in the readme and direct people to the docs in flow. Do you agree @rsolomon? I haven't been following flow's development lately so I'm not sure.

I also remember we had trouble defining HOCs that would play nicely with different combinations of props in the inner and outer components (eg. if the inner component had exact props and the outer had inexact props). Is this problem fixed by `React.ElementConfig`?
